### PR TITLE
Many improvements to AutoSign

### PR DESCRIPTION
**Pull request by RealMuffinTime**
## Type of change

- [ ] Bug fix
- [x] New feature

## Description

- Remove hardcoded time placeholder from MeteorStarscript.java
- Added persistent settings to save sign lines to
- Added Starscript support
- Added possibility to also apply back side of sign
- Added possibility to overwrite or keep existent text on sign
  When not applying anything, open edit screen normally
- 

# How Has This Been Tested?
- Placing many signs, in diffrent situations
- On a local vanilla server.

# Checklist:

- [x] My code follows the style guidelines of this project.
- [x] I have added comments to my code in more complex areas.
- [ ] I have tested the code in both development and production environments.

### CODE REVIEW

**Comment by RacoonDog** [comment]
what was wrong with the time placeholder? it looks correct from looking at the code

**Comment by RealMuffinTime** [comment]
It was not wrong, but it did always overwrite the time placeholder from the Starscript StandardLib.timeFormat object.

**Comment by RacoonDog** [comment]
> It was not wrong, but it did always overwrite the time placeholder from the Starscript StandardLib.timeFormat object.

ahh i see

File: src/main/java/meteordevelopment/meteorclient/systems/modules/world/AutoSign.java [hunk 1]
```diff
@@ -6,42 +6,293 @@
...
+import java.util.Objects;
+import java.util.TimeZone;
 
 public class AutoSign extends Module {
-    private String[] text;
 
-    public AutoSign() {
-        super(Categories.World, "auto-sign", "Automatically writes signs. The first sign's text will be used.");
-    }
+    private boolean applyingBack = false;
```
Comment (RacoonDog) [inline src/main/java/meteordevelopment/meteorclient/systems/modules/world/AutoSign.java:39]: place variables after settings/setting groups

File: src/main/java/meteordevelopment/meteorclient/systems/modules/world/AutoSign.java [hunk 2]
```diff
@@ -6,42 +6,293 @@
...
+
+    private final SettingGroup front = settings.createGroup("Front");
+    private final SettingGroup back = settings.createGroup("Back");
+    private final SettingGroup other = settings.createGroup("Other");
+
+    private final List<Script> frontScripts = new ArrayList<>();
+    private final List<Script> backScripts = new ArrayList<>();
```
Comment (RacoonDog) [inline src/main/java/meteordevelopment/meteorclient/systems/modules/world/AutoSign.java:46]: place variables after settings/setting groups

File: src/main/java/meteordevelopment/meteorclient/systems/modules/world/AutoSign.java [hunk 3]
```diff
@@ -6,42 +6,293 @@
...
+        .description("Enable auto writing front of sign.")
+        .defaultValue(true)
+        .build()
+    );
+
+    private final Setting<String> line1front = front.add(new StringSetting.Builder()
+        .name("front-line-1")
+        .description("Text in first line.")
+        .defaultValue("{player}")
+        .onChanged(strings -> recompile(front, frontScripts, 1))
```
Comment (RacoonDog) [inline src/main/java/meteordevelopment/meteorclient/systems/modules/world/AutoSign.java:60]: instead of passing the `SettingGroup` and getting the string from the index, you can just pass in `strings`

File: src/main/java/meteordevelopment/meteorclient/systems/modules/world/AutoSign.java [hunk 4]
```diff
@@ -6,42 +6,293 @@
...
-    public AutoSign() {
-        super(Categories.World, "auto-sign", "Automatically writes signs. The first sign's text will be used.");
-    }
+    private boolean applyingBack = false;
+
+    private final SettingGroup front = settings.createGroup("Front");
+    private final SettingGroup back = settings.createGroup("Back");
+    private final SettingGroup other = settings.createGroup("Other");
+
+    private final List<Script> frontScripts = new ArrayList<>();
```
Comment (RacoonDog) [inline src/main/java/meteordevelopment/meteorclient/systems/modules/world/AutoSign.java:45]: instead of using `ArrayList`, make fixed size lists using `Arrays.asList(new String[4])`

File: src/main/java/meteordevelopment/meteorclient/systems/modules/world/AutoSign.java [hunk 5]
```diff
@@ -6,42 +6,293 @@
...
+        }
+    }
+
+    private void recompile(SettingGroup side, List<Script> sideScripts, int line) {
+        String compileLine = (String) side.getByIndex(line).get();
+
+        Script script = MeteorStarscript.compile(compileLine);
 
-        event.cancel();
+        if (sideScripts.size() <= line - 1) {
```
Comment (RacoonDog) [inline src/main/java/meteordevelopment/meteorclient/systems/modules/world/AutoSign.java:292]: you can remove this condition with the fixed size lists

File: src/main/java/meteordevelopment/meteorclient/systems/modules/world/AutoSign.java [hunk 6]
```diff
@@ -6,42 +6,293 @@
...
+        if (!(event.screen instanceof AbstractSignEditScreen)) {return;}
+        if (applyingBack) {
+            applyingBack = false;
+            event.cancel();
+            return;
+        }
 
+        // apply custom format and timezone, not the nicest (temporary overwriting global date and time format)
+        try {
+            StandardLib.dateFormat.applyPattern(date.get());
```
Comment (RacoonDog) [inline src/main/java/meteordevelopment/meteorclient/systems/modules/world/AutoSign.java:200]: i would rather the module not modify global state like that, it would be better to modify `MeteorStarscript` to add specialized time & date functions that accept format strings & timezone identifiers as arguments

File: src/main/java/meteordevelopment/meteorclient/systems/modules/world/AutoSign.java [hunk 7]
```diff
@@ -6,42 +6,293 @@
...
+        }
+
+        boolean updated = false;
         SignBlockEntity sign = ((AbstractSignEditScreenAccessor) event.screen).getSign();
 
-        mc.player.networkHandler.sendPacket(new UpdateSignC2SPacket(sign.getPos(), true, text[0], text[1], text[2], text[3]));
+        if (frontEnabled.get()) {
+            boolean written = false;
+            for (Text text : sign.getFrontText().getMessages(false))
+                if (!Objects.equals(text.getString(), "") && !written) {
```
Comment (RacoonDog) [inline src/main/java/meteordevelopment/meteorclient/systems/modules/world/AutoSign.java:216]: `!Objects.equals(text.getString(), "")` can be rewritten as `!text.getString().isEmpty()`

File: src/main/java/meteordevelopment/meteorclient/systems/modules/world/AutoSign.java [hunk 8]
```diff
@@ -6,42 +6,293 @@
...
+            }
+
+            if (offhand_swapped) {
+                mc.player.networkHandler.sendPacket(new PlayerActionC2SPacket(PlayerActionC2SPacket.Action.SWAP_ITEM_WITH_OFFHAND, mc.player.getBlockPos(), mc.player.getFacing()));
+            }
+        }
+
+        if (backEnabled.get()) {
+            boolean written = false;
+            for (Text text : sign.getBackText().getMessages(false))
```
Comment (RacoonDog) [inline src/main/java/meteordevelopment/meteorclient/systems/modules/world/AutoSign.java:259]: please add braces

File: src/main/java/meteordevelopment/meteorclient/systems/modules/world/AutoSign.java [hunk 9]
```diff
@@ -6,42 +6,293 @@
...
+            StandardLib.timeFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
+        }
+
+        boolean updated = false;
         SignBlockEntity sign = ((AbstractSignEditScreenAccessor) event.screen).getSign();
 
-        mc.player.networkHandler.sendPacket(new UpdateSignC2SPacket(sign.getPos(), true, text[0], text[1], text[2], text[3]));
+        if (frontEnabled.get()) {
+            boolean written = false;
+            for (Text text : sign.getFrontText().getMessages(false))
```
Comment (RacoonDog) [inline src/main/java/meteordevelopment/meteorclient/systems/modules/world/AutoSign.java:215]: please add braces

**Review by RacoonDog — CHANGES_REQUESTED** [review]

File: src/main/java/meteordevelopment/meteorclient/systems/modules/world/AutoSign.java (hunk 6 above)
Comment (RealMuffinTime) [reply src/main/java/meteordevelopment/meteorclient/systems/modules/world/AutoSign.java:200]: True.

I saw that there is a [pull requests](https://github.com/MeteorDevelopment/starscript/pull/21) which could resolve this issue.

**Comment by RealMuffinTime** [comment]
While testing I found some bigger issues, need to look into how to solve these.

**Comment by RacoonDog** [comment]
> While testing I found some bigger issues, need to look into how to solve these.

what are the issues?

**Comment by RealMuffinTime** [comment]
First something unrelated: sign.getEditor() is always null for me?

So the module works fine on a server which accepts sign editing (vanilla).
But I also tested it on servers where editing is not possible, this turned out to have negative side effects, while having back sign writing enabled. Double placed signs on 2b2t (2b2t doesn't allow editing), every second sign empty on constantiam (older mc version).

**Comment by RealMuffinTime** [comment]
I tried to use sign.getEditor() to verify if the BlockInteraction was successfull and if the sign could be edited.

**Comment by RealMuffinTime** [comment]
Is checked and tested. Added a warning, if server doesn't support back sign editing.

**Comment by RealMuffinTime** [comment]
Hi, I would really like, to see this merged. :)
Also if https://github.com/MeteorDevelopment/starscript/pull/21 is not yet being merged.

Or is there something else holding up this PR?
